src/watcher.js

```diff
@@ -0,0 +1,88 @@
+require('dotenv').config()
+const path = require('path')
+const Web3 = require('web3')
+const { connectWatcherToQueue } = require('./services/amqpClient')
+const { getBlockNumber } = require('./utils/utils')
+const processDeposits = require('./events/processDeposits')
+const processCollectedSignatures = require('./events/processCollectedSignatures')
+const processWithdraw = require('./events/processWithdraw')
+const { redis } = require('./services/redisClient')
+
+const config = require(path.join('../config/', process.argv[2]))
+
+const provider = new Web3.providers.HttpProvider(config.url)
+const web3Instance = new Web3(provider)
+const bridgeContract = new web3Instance.eth.Contract(config.abi, config.contractAddress)
+const lastBlockRedisKey = `${config.id}:lastProcessedBlock`
+let lastProcessedBlock = 0
+
+async function initialize() {
+  await getLastProcessedBlock()
+  connectWatcherToQueue({
+    queueName: config.queue,
+    cb: runMain
+  })
+}
+
+async function runMain({ sendToQueue, isAmqpConnected }) {
+  if (isAmqpConnected() && redis.status === 'ready') {
+    await main({ sendToQueue })
+  }
+
+  setTimeout(() => {
+    runMain({ sendToQueue, isAmqpConnected })
+  }, 1000)
+}
+
+async function getLastProcessedBlock() {
+  const result = await redis.get(lastBlockRedisKey)
+  lastProcessedBlock = result ? Number(result) : 0
+}
+
+function updateLastProcessedBlock(lastBlockNumber) {
+  lastProcessedBlock = lastBlockNumber
+  redis.set(lastBlockRedisKey, lastProcessedBlock)
+}
+
+function processEvents(events) {
+  switch (config.id) {
+    case 'deposit':
+      return processDeposits(events)
+    case 'collected-signatures':
+      return processCollectedSignatures(events)
+    case 'withdraw':
+      return processWithdraw(events)
+    default:
+      return []
+  }
+}
+
+async function main({ sendToQueue }) {
+  try {
+    const lastBlockNumber = await getBlockNumber(web3Instance)
+    if (lastBlockNumber === undefined || lastBlockNumber === lastProcessedBlock) {
```

Comment:
In what circumstances is `lastBlockNumber` undefined?

Reply:
With this new code I think there is no use case for that `undefined` check, I will remove it.